designs/NuGetExtensibility/NuGetExtensibilityServices.md

```diff
@@ -0,0 +1,161 @@
+# NuGet extensibility services
+
+* Status: **Reviewed**
+* Author(s): [Nikolche Kolev](https://github.com/nkolev92)
+* Issue: [8896](https://github.com/NuGet/Home/issues/9062) Guidelines for asynchronous NuGet extensibility services
+* Type: Architecture
+
+## Problem Background
+
+NuGet has a set of IVs* APIs such as IVsPackageInstaller, IVsPackageUninstaller, IVsPackageSourceProvider etc.
+These APIs were mostly designed and implemented during the Dev12-14 days when packages.config was style of choice.
+Nearly all of these APIs are synchronous and are defined in an [embeddable](https://www.nuget.org/packages/NuGet.VisualStudio/) assembly.
+
+According to the VS threading cookbook and [analyzers](https://github.com/microsoft/vs-threading/blob/master/doc/analyzers/VSTHRD010.md), most IVs synchronous services have threading requirements and should be invoked on the UI thread.
+As such a lot of our customers as a pattern invoke the NuGet APIs from the UI thread.
+
+This leads to a lot of performance and reliability issues with NuGet as the culprit.
+
+Despite being synchronous IVs* some of our APIs can be invoked off the UI thread, but that's not consistently documented.
+Given that most of our code is already asynchronous, there's no need why we cannot expose services that are asynchronous.
+
+## Who are the customers
+
+All Visual Studio users wil reap the benefits.
+The primary users of these new APIs will be extension authors and Visual Studio components like Roslyn.
+
+## Requirements
+
+* Asynchronous and performant NuGet extensibility services
+* Ability to invoke these through RPC. These days there are components that are running out of devenv.exe
+
+## Goals
+
+* Define a guideline for designing asynchronous NuGet services
+* Introduce a new NuGet.VisualStudio.Contracts
+* Expose them through an extensibility mechanism to allow out of process invocation
+
+## Non-Goals
+
+* Define the APIs for all the services that will be migrated.

```

Comment:
Will we have one-on-one equivalent with current NuGet IVs* interfaces?

Reply:
This just covers the guidelines, not an implementation of a specific service. 
As such the goal of this architecture doc is not to define the services. 

All of the services that need migration will be evaluated at their own merits.